lib/Yield/Support/GraphLayout/SugiyamaStyle/GraphPreparation.cpp

```diff
@@ -42,39 +43,66 @@ static SelfLoopContainer extractSelfLoops(InternalGraph &Graph) {
-    // If there's more than one, add a new virtual node with all the real
-    // entry nodes as its direct successors.
-
-    // BUT if the currently set entry node is already virtual, remove it first:
-    // this prevents the possibility of chaining virtual entry nodes when this
-    // function is invoked on a slightly-modified graph multiple times.
-    if (Graph.getEntryNode() != nullptr) {
-      if (Graph.getEntryNode()->IsVirtual) {
-        Graph.removeNode(Graph.getEntryNode());
-        if (MaybeRanks != nullptr)
-          MaybeRanks->erase(Graph.getEntryNode());
+/// Adds a virtual node to the graph to serve as an entry point, while
+/// also cleanly removing it when this object is being destroyed.
+class VirtualEntryPointGuard {
+private:
+  InternalGraph &Graph;
+  InternalNode &Node;
+  RankContainer *MaybeRanks;
+
+  VirtualEntryPointGuard(const VirtualEntryPointGuard &) = delete;
+  VirtualEntryPointGuard(VirtualEntryPointGuard &&) = delete;
+  VirtualEntryPointGuard &operator=(const VirtualEntryPointGuard &) = delete;
+  VirtualEntryPointGuard &operator=(VirtualEntryPointGuard &&) = delete;
+
+public:
+  explicit VirtualEntryPointGuard(InternalGraph &Graph,
+                                  RankContainer *MaybeRanks = nullptr) :
+    Graph(Graph), Node(*Graph.makeVirtualNode()), MaybeRanks(MaybeRanks) {
+
+    std::vector<InternalNode *> EntryPoints = entryPoints(&Graph);
+    revng_assert(!EntryPoints.empty());
+
+    for (InternalNode *EntryPoint : EntryPoints) {
+      if (EntryPoint != &Node) {
+        revng_assert(!Node.hasSuccessor(EntryPoint));
+        Node.addSuccessor(EntryPoint, InternalNode::Edge(Graph));
       }
     }
 
-    InternalNode *EntryPoint = Graph.makeVirtualNode();
-    for (InternalNode *Node : Graph.nodes())
-      if (!Node->hasPredecessors() && Node->index() != EntryPoint->index())
-        EntryPoint->addSuccessor(Node, InternalNode::Edge(Graph));
-    Graph.setEntryNode(EntryPoint);
+    revng_assert(Graph.getEntryNode() == nullptr);
+    Graph.setEntryNode(&Node);
   }
-}
+
+  InternalNode *operator*() { return &Node; }
+  InternalNode *operator*() const { return &Node; }
+  InternalNode *operator->() { return &Node; }
+  InternalNode *operator->() const { return &Node; }
+
+  ~VirtualEntryPointGuard() {
+    revng_assert(Node.IsVirtual == true);
+    revng_assert(Graph.getEntryNode() == &Node);
+    Graph.setEntryNode(nullptr);
+
+    static constexpr auto Filter = [](InternalNode *const &,
+                                      InternalNode *const &Successor) {
+      revng_assert(Successor != nullptr);
+      return Successor->IsVirtual;
+    };
+    using FilteredGraphType = NodePairFilteredGraph<InternalNode *, Filter>;
+
+    std::vector<NodeView> ToRemove;
+    for (InternalNode *Current : llvm::depth_first(FilteredGraphType(&Node)))
+      ToRemove.emplace_back(Current);
+
+    for (NodeView Node : ToRemove) {
+      if (MaybeRanks != nullptr)
+        MaybeRanks->erase(Node);
```

Comment:
Then some ranks might become empty. Might this be a problem?

Reply:
But, said that, to answer the original question, when `Ranks` get converted into the `Layers`, there's an extensive clean-up to ensure the "problem" you are worrying about here can never happen (again).

Reply:
Ok cool thanks.